kolibri/core/discovery/utils/network/search.py

```diff
@@ -0,0 +1,174 @@
+            try:
+                info = ServiceInfo(
+                    SERVICE_TYPE,
+                    name=".".join([id, SERVICE_TYPE]),
+                    server=".".join([id, LOCAL_DOMAIN, ""]),
+                    address=USE_IP_OF_OUTGOING_INTERFACE,
+                    port=self.port,
+                    properties=self.data,
+                )
+
+                ZEROCONF_STATE["zeroconf"].register_service(info, ttl=60)
+
+                self.info = info
+
+            except NonUniqueNameException:
+                # if there's a name conflict, append incrementing integer until no conflict
+                i += 1
+                id = "%s-%d" % (self.id, i)
+
+            if i > 100:
+                raise NonUniqueNameException()
+
+        self.id = id
+
+        return self
+
+    def unregister(self):
+
+        assert self.info is not None, "Service is not registered!"
+
+        ZEROCONF_STATE["zeroconf"].unregister_service(self.info)
+
+        self.info = None
+
+    def cleanup(self, *args, **kwargs):
+
+        if self.info and ZEROCONF_STATE["zeroconf"]:
+            self.unregister()
+
+
+class KolibriZeroconfListener(object):
+
+    instances = {}
+
+    def add_service(self, zeroconf, type, name):
+        info = zeroconf.get_service_info(type, name)
+        id = _id_from_name(name)
+        ip = socket.inet_ntoa(info.address)
+        self.instances[id] = {
+            "id": id,
+            "ip": ip,
+            "local": ip in get_all_addresses(),
+            "port": info.port,
+            "host": info.server.strip("."),
+            "data": {key: json.loads(val) for (key, val) in info.properties.items()},
+        }
+        logger.info(
+            "Kolibri instance '%s' joined zeroconf network; service info: %s\n"
+            % (id, self.instances[id])
+        )
+
+    def remove_service(self, zeroconf, type, name):
+        id = _id_from_name(name)
+        logger.info("\nKolibri instance '%s' has left the zeroconf network.\n" % (id,))
+        if id in self.instances:
+            del self.instances[id]
+
+
+def get_available_instances(timeout=2, include_local=True):
+    """Retrieve a list of dicts with information about the discovered Kolibri instances on the local network,
+    filtering out those that can't be accessed at the specified port (via attempting to open a socket)."""
+    if not ZEROCONF_STATE["listener"]:
+        initialize_zeroconf_listener()
+        time.sleep(3)
+    instances = []
+    for instance in ZEROCONF_STATE["listener"].instances.values():
+        if instance["local"] and not include_local:
+            continue
+        if not _is_port_open(instance["ip"], instance["port"], timeout=timeout):
+            continue
+        instance["self"] = (
+            ZEROCONF_STATE["service"] and ZEROCONF_STATE["service"].id == instance["id"]
+        )
+        instances.append(instance)
+    return instances
+
+
+def register_zeroconf_service(port, id):
```

Comment:
Why does this function need the port as a parameter?
Is not possible to get it from the options or reading it from the `PID_FILE` as it is done in other parts of the code?
This `port` option pushes changes in many parts of this PR code that might not be needed.

Reply:
Ah, PID_FILE could be a good place, yes. I originally wanted to pull it from `conf.OPTIONS`, except it can be overridden with `--port=XYZ`, which doesn't get stored anywhere (except in the PID, I guess), just passed in to the start command function as an argument, which was the pattern I followed.

Reply:
Yes, I understand your concerns but, as I understand the code, the value in the PID file is affected both by conf.OPTIONS and env variables, thus it should be the real value for the port. Taking it from there would avoid many modifications you had to do just to drag the port parameter here.

Reply:
Yep, agreed! Can't see any issues with that approach. Will update.

The PID is also written into there for kolibri-server, yes? Based on what uwsgi/nginx are exposing?

Reply:
kolibri-server  takes the value from conf.OPTIONS because there's no warranty the part of code in Kolibri that saves the PID has been executed before kolibri-server starts. The env variable option does not have sense when running it as a daemon.
After kolibri finishes the starting process, the PID file will have the same value that is in conf.OPTIONS so I think we can consider we are safe on this side too.